### ISSUE TITLE
fr: support `# syntax`

### ISSUE THREAD

**Issue by lizelive** [containers/buildah#4211]
buildah does not support [alt sytntax](https://docs.docker.com/engine/reference/builder/#syntax)
one way to fix is to implement the [llb](https://github.com/moby/buildkit/blob/master/solver/pb/ops.proto)

this issue is more general case of https://github.com/containers/buildah/issues/3497

**Comment by flouthoc** [comment]
Hi @lizelive , Could you please explain objective of this RFE in detail please, do you want to use `buildkit-nix` but it invokes buildah underneath ?

**Comment by flouthoc** [comment]
> Hi @lizelive , Could you please explain objective of this RFE in detail please, do you want to use `buildkit-nix` but it invokes buildah underneath ?

@lizelive friendly ping.

**Comment by rhatdan** [comment]
@AkihiroSuda WDYT?

**Comment by AkihiroSuda** [comment]
> @AkihiroSuda WDYT?

Sounds good, but it is likely to require replacing almost entire part of Buildah with BuildKit.

**Comment by lizelive** [comment]
> Hi @lizelive , Could you please explain objective of this RFE in detail please, do you want to use `buildkit-nix` but it invokes buildah underneath ?

yeah and similar. it seems that it would reduce the amount of code required to be maintained because the lib is similar to buildah

**Comment by rhatdan** [comment]
@flouthoc We need to discuss this one.

**Comment by g-suraj** [comment]
Hey folks, I see this issue has not been updated in a while, are there any updates on this?

I'm currently working on migrating a large container-based CI system away from Docker to Podman, which largely manages building images and running containerised tests. This has had quite stellar results so far, but the lack of buildkit support is a blocker for us in terms of fully adopting it. Are there any plans to support buildkit as an alternative backend in `buildah`? If not, would you be open to such a contribution to `buildah`?

**Comment by rhatdan** [comment]
We have no one to work on this now, but we are a fully open project and consider all PRs.

**Comment by BenjaminSchubert** [comment]
Hey!

I'd be interested by contributing to this feature. I however am not sure where to start.

@rhatdan, in Feb 2023 you mentioned:

> @flouthoc We need to discuss this one.

I was wondering if some discussions already took place about hot to implement it or if you (or other maintainers) were expecting it to be implemented in a specific way? If there's additional context not present in this issue, please let me know.

From reading this issue, what I get is that there needs to be a decision on how deeply to integrate with buildkit. I can see at least three solutions (in no particular order):

1. Use buildkit as a library to build most images where compatible, and remove the duplication from inside buildah
2. Implement the llb spec fully inside buildah
3. Keep buildah as is, and use the buildkit library only when detecting an input file starting with `# syntax`

What would you think is the best way to start? (And is this the best medium or should this be brought to the mailing list?)

**Comment by rhatdan** [comment]
We currently ignore the # syntax and attempt to stick to buildkit syntax in general.  We  think this is the future and see little value in supporting the older format.

Is there specific things that we do not do in BuildKit format that you would like to see added?

In some cases Buildah is ahead of support Docker, for example we recently added support for COPY --exclude ... flag.

**Comment by BenjaminSchubert** [comment]
The feature we are most interested in is the ability to specify custom frontends. In our specific case, we have a custom buildkit frontend that is responsible for generating the instructions (or part of them) to build the image.

Something similar to https://github.com/EricHripko/buildkit-fdk.

This pattern has allowed us to greatly reduce the complexity of building images, and helped us increase consistency across them

**Comment by BenjaminSchubert** [comment]
@rhatdan sorry to ping again, does what I mentioned above make sense? Would you be open in having something similar in buildah? If not, is there any recommendation you would have for such a use-case?

**Comment by folliehiyuki** [comment]
I want to use `buildah` to build Unikraft unikernels using the frontend syntax from https://github.com/unikraft/kraftkit/tree/staging/tools/dockerfile-llb-frontend. It'd be great if this feature is supported.
